### PR TITLE
Add tests for table component

### PR DESCRIPTION
**Pull request by igloosi**
We break this down into groups (as represented by describe blocks).

Trello https://trello.com/c/mdqq5RnP/645-automated-tests-for-table-component

### CODE REVIEW

**Comment by 36degrees** [comment]
I think I would agree with Nick here. I find it really hard to follow the tests and to understand what the intent is. I think there's as much chance of introducing bugs in the logic of the tests as there is in having bugs in the code we're trying to test.

https://github.com/mawrkus/js-unit-testing-guide#avoid-logic-in-your-tests

**Comment by igloosi** [comment]
was hoping to cut down on the number of loops for the default example for each cell and row, when checking data

**Comment by nickcolley** [comment]
@igloosi I suppose since we only have three rows it'd look like

```
loop 3
  assertion
endloop
```

```
assertion
assertion
assertion
```

So the compromise is verbose tests, which for me personally is OK since looping behaviours are harder to grok and easier to hide bad assertions?

**Comment by igloosi** [comment]
cool, i'll rewrite for more verbosity

**Comment by 36degrees** [comment]
How do we feel about the tests asserting against their own inputs?

i.e. doing things like
```
args = { "text": "bar" }
result = foo(args)

expect(result.text).toEqual(args.text)
```

rather than

```
result = doSomething({"text": "bar"})

expect(result.text).toEqual("bar")
```

I would generally prefer the latter, as I think: 

- it makes the intention clearer (as someone reading the tests, you don't have to follow and parse variable declarations to understand what we want to be happening)
- it is a lot harder to introduce bugs into your test
- it's how we're doing things in all (?) the other test cases so far

File: src/components/table/template.test.js [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+          }
+        ]
+      ]
+    }
+    const { $ } = render('table', args)
+    const $component = $('.govuk-c-table')
+    const $componentBody = $component.children('tbody')
+    const $componentBodyRow = $componentBody.children('tr')
+
+    expect(parseInt($componentBodyRow.eq(0).children().eq(1).attr('colspan'), 10)).toBe(args.rows[0][1].colspan)
```
Comment (36degrees) [inline src/components/table/template.test.js]: If we used `toEqual` here, could we get rid of the casting?

**Comment by igloosi** [comment]
will revise

File: src/components/table/template.test.js (hunk 1 above)
Comment (igloosi) [reply src/components/table/template.test.js]: hmmm, with toEqual:

 Expected value to equal:
      2
 Received:
     "2"

File: src/components/table/template.test.js [hunk 2]
```diff
@@ -0,0 +1,362 @@
...
+      expect($tableHead).toHaveLength(1)
+      expect($tableHeadRow).toHaveLength(1)
+      expect($tableHeadCell).toHaveLength(3)
+    })
+
+    it('renders a caption', () => {
+      const { $ } = render('table', examples['table-with-caption-and-head'])
+      const $caption = $('.govuk-c-table__caption')
+
+      expect($caption).toHaveLength(1)
```
Comment (nickcolley) [inline src/components/table/template.test.js]: I don't think we need this assertion since the next one down will fail if the caption does not exist.

File: src/components/table/template.test.js [hunk 3]
```diff
@@ -0,0 +1,362 @@
...
+    it('renders correct text in cell 2 of row 1', () => {
+      expect($tableRow.eq(0).find($tableCell).eq(1).text()).toEqual('£85')
+    })
+
+    it('renders correct text in cell 3 of row 1', () => {
+      expect($tableRow.eq(0).find($tableCell).eq(2).text()).toEqual('£95')
+    })
+  })
+
+  describe('variant', () => {
```
Comment (nickcolley) [inline src/components/table/template.test.js]: Does this need to be in a describe block? What is it describing?

File: src/components/table/template.test.js [hunk 4]
```diff
@@ -0,0 +1,362 @@
...
+            'format': 'numeric'
+          }
+        ]
+      ]
+    })
+    const $component = $('.govuk-c-table')
+    const $tableRow = $component.find('.govuk-c-table .govuk-c-table__body .govuk-c-table__row')
+
+    expect($tableRow.eq(0).children().eq(0).get(0).tagName).toEqual('th')
+    expect($tableRow.eq(0).children().eq(0).hasClass('govuk-c-table__header')).toBeTruthy()
```
Comment (nickcolley) [inline src/components/table/template.test.js]: We could swap this around and make this a bit easier to read by doing

Instead of 


```js
expect($tableRow.eq(0).children().eq(0).get(0).tagName).toEqual('th')
 expect($tableRow.eq(0).children().eq(0).hasClass('govuk-c-table__header')).toBeTruthy()
```

try

```js
const $tableBody = $component.find('.govuk-c-table .govuk-c-table__body')

const $firstTableRow = $tableBody.find('.govuk-c-table__row:first-child')
const $firstTableHeader = $firstTableRow.find('govuk-c-table__header:first-child')
expect($firstTableHeader.get(0).tagName).toEqual('th')

const $lastTableRow = $tableBody.find('.govuk-c-table__row:last-child')
const $lastTableHeader = $lastTableRow.find('govuk-c-table__header:last-child')
expect($lastTableHeader.get(0).tagName).toEqual('th')

```

File: src/components/table/template.test.js (hunk 3 above)
Comment (alex-ju) [inline src/components/table/template.test.js]: could we get these two tests outside 'variant', even if they're using examples from yaml. also, and I know I'm being picky with this one, but can we order them somehow in a similar way we have them in the table or arguments (classes, caption..) would make it easier to follow and make sure we cover all the arguments.

Comment (igloosi) [reply src/components/table/template.test.js]: removed

**Comment by igloosi** [comment]
@nickcolley  and @alex-ju  update based on comments

File: src/components/table/template.test.js [hunk 5]
```diff
@@ -0,0 +1,374 @@
+/* globals describe, it, expect */
+
+const { render, getExamples } = require('../../../lib/jest-helpers')
+
+const examples = getExamples('table')
+describe('Table', () => {
+
```
Comment (alex-ju) [inline src/components/table/template.test.js]: _Block must not be padded by blank lines_

File: src/components/table/template.test.js [hunk 6]
```diff
@@ -0,0 +1,374 @@
...
+    })
+    const $component = $('.govuk-c-table')
+    const $tableBody = $component.find('.govuk-c-table .govuk-c-table__body')
+    const $firstTableRow = $tableBody.find('.govuk-c-table__row:first-child')
+    const $firstTableHeader = $firstTableRow.find('.govuk-c-table__header:first-child')
+
+    expect($firstTableHeader.get(0).tagName).toEqual('th')
+    expect($firstTableHeader.attr('scope')).toEqual('row')
+    expect($firstTableHeader.hasClass('govuk-c-table__header')).toBeTruthy()
+
```
Comment (alex-ju) [inline src/components/table/template.test.js]: _More than one blank line not allowed_

File: src/components/table/template.test.js [hunk 7]
```diff
@@ -0,0 +1,374 @@
...
+
+    const $lastTableRow = $tableBody.find('.govuk-c-table__row:last-child')
+    const $lastTableHeader = $lastTableRow.find('.govuk-c-table__header:first-child')
+
+    expect($lastTableHeader.get(0).tagName).toEqual('th')
+    expect($lastTableHeader.attr('scope')).toEqual('row')
+    expect($lastTableHeader.hasClass('govuk-c-table__header')).toBeTruthy()
+
```
Comment (alex-ju) [inline src/components/table/template.test.js]: _Block must not be padded by blank lines_

File: src/components/table/template.test.js (hunk 5 above)
Comment (igloosi) [reply src/components/table/template.test.js]: fixed

File: src/components/table/template.test.js (hunk 7 above)
Comment (igloosi) [reply src/components/table/template.test.js]: fixed
